cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CruiseControlReconciler.java

```diff
@@ -247,52 +246,74 @@ protected Future<Void> certificatesSecret(Clock clock) {
+                topicOperatorManagedApiSecretIndex = futures.size() - 1;
+            } else {
+                topicOperatorManagedApiSecretIndex = null;
+            }
+
+            return Future.join(futures)
+                .compose(
                     compositeFuture -> {
-                        Secret oldSecret = compositeFuture.resultAt(0);
-                        Secret topicOperatorApiSecret = compositeFuture.resultAt(1);
-                        String cruiseControlAuthFile = oldSecret != null ? Util.decodeFromBase64(oldSecret.getData().get(API_AUTH_FILE_KEY)) : null;
-                        CruiseControl.CruiseControlUser toAdminUser = topicOperatorApiSecret != null
-                            ? new CruiseControl.CruiseControlUser(Util.decodeFromBase64(topicOperatorApiSecret.getData().get(API_TO_ADMIN_NAME_KEY)),
-                            Util.decodeFromBase64(topicOperatorApiSecret.getData().get(API_TO_ADMIN_PASSWORD_KEY)))
-                            : null;
-                        // generate a new CC API secret if there is no CC auth file, or there is no TO API secret, or TO admin password changed
-                        Secret newSecret = cruiseControlAuthFile == null || toAdminUser == null || !cruiseControlAuthFile.contains(toAdminUser.password())
-                            ? cruiseControl.generateApiSecret(passwordGenerator, null, toAdminUser)
-                            : cruiseControl.generateApiSecret(passwordGenerator, oldSecret, null);
-                        this.apiSecretHash = ReconcilerUtils.hashSecretContent(newSecret);
-                        return secretOperator.reconcile(reconciliation, reconciliation.namespace(), CruiseControlResources.apiSecretName(reconciliation.name()), newSecret)
+                        Secret oldCentralizedApiSecret = compositeFuture.resultAt(centralizedApiSecretIndex);
+                        Secret userManagedApiSecret = userManagedApiSecretIndex != null ? compositeFuture.resultAt(userManagedApiSecretIndex) : null;
+                        Secret topicOperatorManagedApiSecret = topicOperatorManagedApiSecretIndex != null ? compositeFuture.resultAt(topicOperatorManagedApiSecretIndex) : null;
+
+                        Map<String, ApiCredentials.Entry> apiCredentials = new HashMap<>();
+                        apiCredentials.putAll(ApiCredentials.generateCoManagedApiCredentials(passwordGenerator, oldCentralizedApiSecret));
+                        apiCredentials.putAll(ApiCredentials.generateUserManagedApiCredentials(userManagedApiSecret, cruiseControl.getUserManagedApiSecretKey()));
+                        apiCredentials.putAll(ApiCredentials.generateToManagedApiCredentials(topicOperatorManagedApiSecret));
+
+                        Map<String, String> mapWithApiCredentials = ApiCredentials.generateMapWithApiCredentials(apiCredentials);
+                        Secret newCentralizedApiUsersSecret = cruiseControl.generateApiSecret(mapWithApiCredentials);
```

Comment:
As I commented on the utility class ... this should be maybe wrapped into one public method there?